### PR TITLE
Integrating gpu exploration app

### PR DESCRIPTION
**Pull request by FoodChain1028**
The gpu exploration app is made using `mopro init`. This app explores the msm functions that we implements in `mopro-core` and test in rust, and can be run on simulator and mobile.

### CODE REVIEW

File: research/gpu-exploration-app/README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+> please install `mopro-cli` by following the instructions [here](https://github.com/zkmopro/mopro/tree/main/mopro-cli)
+
+### Prepare
+
+`mopro prepare`
+
+### Build
+
+`mopro build --platform ios`
```
Comment (vivianjeng) [inline research/gpu-exploration-app/README.md:19]: The command should be updated
```suggestion
`mopro build --platforms ios`
```

File: research/gpu-exploration-app/README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+### Test
+
+`mopro test`
+
+The result would be shown as the following and in the folder `core/benchmarks/results/`
```
Comment (vivianjeng) [inline research/gpu-exploration-app/README.md:25]: The `mopro test` doesn't work for me.
maybe with clearer instructions how to reproduce this?

errors:
```sh
error[E0277]: the trait bound `ark_ff::fields::models::Fp256<FrParameters>: ark_ff::PrimeField` is not satisfied
  --> /Users/zhengyawen/.cargo/git/checkouts/mopro-471ecc433f4d4aff/371bfe5/mopro-core/src/middleware/gpu_explorations/arkworks_pippenger.rs:9:1
   |
9  | / pub fn benchmark_msm<I>(
10 | |     instances: I,
11 | |     iterations: u32,
12 | | ) -> Result<Vec<Duration>, preprocess::HarnessError>
13 | | where
14 | |     I: Iterator<Item = preprocess::Instance>,
   | |_____________________________________________^ the trait `ark_ff::PrimeField` is not implemented for `ark_ff::fields::models::Fp256<FrParameters>`
...
```

File: research/gpu-exploration-app/ios/ExampleApp/Frameworks/MoproBindings.podspec [hunk 3]
```diff
@@ -0,0 +1,11 @@
+Pod::Spec.new do |spec|
+  spec.name         = 'MoproBindings'
+  spec.version      = '0.1.0'
+  spec.summary      = 'MoproBindings XCFramework'
+  spec.homepage     = 'https://github.com/oskarth/mopro'
```
Comment (vivianjeng) [inline research/gpu-exploration-app/ios/ExampleApp/Frameworks/MoproBindings.podspec:5]: I just fixed the path in templates/mopro-example-app
```suggestion
  spec.homepage     = 'https://github.com/zkmopro/mopro'
```

File: research/gpu-exploration-app/ios/ExampleApp/Frameworks/CircuitBindings.podspec [hunk 4]
```diff
@@ -0,0 +1,12 @@
+Pod::Spec.new do |spec|
+    spec.name         = 'CircuitBindings'
+    spec.version      = '0.1.0'
+    spec.summary      = 'CircuitBindings XCFramework'
+    spec.homepage     = 'https://github.com/oskarth/mopro'
```
Comment (vivianjeng) [inline research/gpu-exploration-app/ios/ExampleApp/Frameworks/CircuitBindings.podspec:5]:
```suggestion
    spec.homepage     = 'https://github.com/zkmopro/mopro'
```

File: research/gpu-exploration-app/core/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+[dependencies]
+num-bigint = { version = "=0.4.3", default-features = false, features = [
+    "rand",
+] }
+ark-bn254 = { version = "=0.4.0" }
+ark-serialize = { version = "=0.4.1", features = ["derive"] }
+
+# FIXME: This doesn't work due to custom build command in mopro-core
+mopro-core = { git = "https://github.com/oskarth/mopro.git", package = "mopro-core", features = ["gpu-benchmarks"] }
```
Comment (vivianjeng) [inline research/gpu-exploration-app/core/Cargo.toml:17]:
```suggestion
mopro-core = { git = "https://github.com/zkmopro/mopro.git", package = "mopro-core", features = ["gpu-benchmarks"] }
```

**Review by vivianjeng — CHANGES_REQUESTED** [review]
Looks good to me.
I think there are two things can be improved
1. remove the example circuits (keccak256) if it is not used (but I think you will need a default circuit (multiplier) to run `mopro build`)
2. put gpu-benchmark in `config.toml`, so it can be easily switched
   (can be in the next or following PR)

**Comment by vivianjeng** [comment]
Can you update the mopro-cli and check your `uniffi-bindgen` version?
I got diffs in `mopro.swift`

File: research/gpu-exploration-app/README.md (hunk 2 above)
Comment (FoodChain1028) [reply research/gpu-exploration-app/README.md:25]: well... this error seems like the cargo did not patch for arkworks version, but I think the Cargo.toml in this folder should have solved this :(

https://github.com/zkmopro/mopro/blob/8cf02da863f1fbe85f09e79964e181ff600afaea/research/gpu-exploration-app/core/Cargo.toml#L30-L35

**Comment by FoodChain1028** [comment]
> Can you update the mopro-cli and check your `uniffi-bindgen` version? I got diffs in `mopro.swift`

Hi @vivianjeng, my current uniffi-bindgen version is 0.25.3 and I updated mopro-cli now. Can you check again?

File: research/gpu-exploration-app/mopro-config.toml [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ptau = "02"                       # ptau to use for trusted setup of circuit, "01" to "27"
+
+# keccak256
+#dir = "core/circuits/keccak256"
+#name = "keccak256_256_test"
+#ptau = "19"
+
+[dylib]
+# NOTE: Dylib support is experimental and requires some fiddling in iOS
+# See https://github.com/oskarth/mopro/pull/37 and https://github.com/oskarth/mopro/pull/38
```
Comment (vivianjeng) [inline research/gpu-exploration-app/mopro-config.toml:26]:
```suggestion
# See https://github.com/zkmopro/mopro/pull/37 and https://github.com/zkmopro/mopro/pull/38
```

File: research/gpu-exploration-app/mopro-config.toml [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+use_dylib = false        # Options: true, false
+name = "keccak256.dylib" # Name of the dylib file, only used if use_dylib is true
+
+[witness]
+# Note: circom-witness-rs is experimental
+# See https://github.com/oskarth/mopro/issues/32 for updates
```
Comment (vivianjeng) [inline research/gpu-exploration-app/mopro-config.toml:32]:
```suggestion
# See https://github.com/zkmopro/mopro/issues/32 for updates
```

**Comment by vivianjeng** [comment]
the `mopro test` works for me. Thank you.
can you check your `uniffi-bindgen --version` in the terminal?
it is still different.

**Comment by vivianjeng** [comment]
@FoodChain1028 yes
mine is also 0.25.3
does it happen to you if you run `mopro build --platforms ios` in the gpu-benchmark dir?

**Comment by FoodChain1028** [comment]
@vivianjeng  
I checked with the binding file generated from mopro and found that the difference. And no, after I ran `mopro build --platforms ios`, the result came out was the one in `gpu-exploration-app` (without keyword `file`).

**Pull request merged**